### PR TITLE
enable backwards compat feature: only contracts after migration are charged for ice creation NTRN-135

### PR DESCRIPTION
**Pull request by quasisamurai**
[TASK](https://hadronlabs.atlassian.net/browse/NTRN-135)

We need changes in ICTX module related to register_fee backwards compatible.

In the upgrade handler of Neutron v1.1.0 we must take the last uploaded code_id of wasm contracts and save it to the storage.

We make register_fee field optional.

When a contract calls RegisterInterchainAccount method, Neutron compares code_id of the calling contract with the saved one:

If it’s less or equal to the saved one - we allow contract to not pay fees for the ICA registration;

if code_id is greater - register_fee is required.

- register_fee is optional field in Neutron CW SDK and in the RegisterInterchainAccount message;

- last code_id is saved into the storage in the upgrade handler for v1.1.0;

- contracts with code_id less or equal then the saved one do not pay fees for ICA registration;

- contract with code_id greater then the saved one pay fees for ICA registration;

- documentation is updated;

- tests are implemented;

### CODE REVIEW

File: x/interchaintxs/keeper/keeper.go [hunk 1]
```diff
@@ -92,3 +92,19 @@ func (k Keeper) ChargeFee(ctx sdk.Context, payer sdk.AccAddress, fee sdk.Coins)
...
+	bytes := store.Get(types.LastCodeIdBeforeUpgrade)
+
+	if bytes == nil {
+		k.Logger(ctx).Debug("Last code id key don't exists, GetLastCodeID returns 0")
+		return 0
+	}
+	return sdk.BigEndianToUint64(bytes)
+}
+
+func (k Keeper) SetLastCodeIDBeforeUpgrade(ctx sdk.Context, id uint64) {
```
Comment (mmikev) [inline x/interchaintxs/keeper/keeper.go]: Unused method

File: app/upgrades/nextupgrade/upgrades.go [hunk 2]
```diff
@@ -238,6 +240,13 @@ func setInterchainTxsParams(ctx sdk.Context, paramsKeepers paramskeeper.Keeper,
 
 	bz := codec.MustMarshal(&currParams)
 	store.Set(interchaintxstypes.ParamsKey, bz)
+
+	wasmStore := ctx.KVStore(wasmStoreKey)
+	bzWasm := wasmStore.Get(wasmtypes.KeyLastCodeID)
+	if bzWasm == nil {
+		return fmt.Errorf("last code ID not found during the upgrade")
+	}
```
Comment (mmikev) [inline app/upgrades/nextupgrade/upgrades.go]: There is no public method in wasmd keeper to use instead of this raw work with the storage?

File: app/upgrades/nextupgrade/upgrades_test.go [hunk 3]
```diff
@@ -159,3 +163,56 @@ func (suite *UpgradeTestSuite) TestAdminModuleUpgrade() {
...
+
+	// register with fees
+	jsonStringAfterUpgrade := `{"register": {"connection_id":"connection-1","interchain_account_id":"test-3"}}`
+	byteEncodedMsgAfterUpgrade := []byte(jsonStringAfterUpgrade)
+	_, err = contractKeeper.Execute(ctx, contractAddressAfterUpgrade, sdk.AccAddress("neutron1weweewe"), byteEncodedMsgAfterUpgrade, sdk.NewCoins(sdk.NewCoin("untrn", sdk.NewInt(1000))))
+	suite.Require().NoError(err)
+
+	// failed register due lack of fees (fees required)
+	_, err = contractKeeper.Execute(ctx, contractAddressAfterUpgrade, sdk.AccAddress("neutron1weweewe"), byteEncodedMsgAfterUpgrade, nil)
+	suite.Error(err)
```
Comment (mmikev) [inline app/upgrades/nextupgrade/upgrades_test.go]: Can we check here that we get exactly fee related error?

File: x/interchaintxs/keeper/msg_server.go [hunk 4]
```diff
@@ -47,8 +47,10 @@ func (k Keeper) RegisterInterchainAccount(goCtx context.Context, msg *ictxtypes.
 		return nil, errors.Wrapf(ictxtypes.ErrNotContract, "%s is not a contract address", msg.FromAddress)
 	}
 
-	if err := k.ChargeFee(ctx, senderAddr, msg.RegisterFee); err != nil {
-		return nil, errors.Wrapf(err, "failed to charge fees to pay for RegisterInterchainAccount msg: %s", msg)
+	if k.sudoKeeper.GetContractInfo(ctx, senderAddr).CodeID >= k.GetLastCodeIDBeforeUpgrade(ctx) {
```
Comment (mmikev) [inline x/interchaintxs/keeper/msg_server.go]: Let's leave a small comment here why we do this

File: x/interchaintxs/types/keys.go [hunk 5]
```diff
@@ -18,7 +18,9 @@ const (
 )
 
 const (
-	prefixParamsKey = iota + 1
+	prefixParamsKey               = iota + 1
+	prefixLastCodeIdBeforeUpgrade = iota + 2
```
Comment (mmikev) [inline x/interchaintxs/types/keys.go]: The same here, a little comment to describe the purpose of the contstant would be very nice in the future

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: x/contractmanager/types/expected_keepers.go [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
 
 import (
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 // WasmKeeper defines the expected interface needed to cam cosmwasm contracts.
 type WasmKeeper interface {
 	HasContractInfo(ctx sdk.Context, contractAddress sdk.AccAddress) bool
+	GetContractInfo(ctx sdk.Context, contractAddress sdk.AccAddress) *wasmtypes.ContractInfo
```
Comment (sotnikov-s) [inline x/contractmanager/types/expected_keepers.go:11]: unused method

File: x/interchaintxs/keeper/keeper.go [hunk 7]
```diff
@@ -92,3 +92,14 @@ func (k Keeper) ChargeFee(ctx sdk.Context, payer sdk.AccAddress, fee sdk.Coins)
 func (k Keeper) GetAuthority() string {
 	return k.authority
 }
+
+func (k Keeper) GetLastCodeIDBeforeUpgrade(ctx sdk.Context) (codeID uint64) {
```
Comment (sotnikov-s) [inline x/interchaintxs/keeper/keeper.go]: I don't like the naming. which upgrade? what if we have more than one upgrade with some code ID? why do we even mention some upgrade in module keeper's interface? let's name it something like `GetLastFreeRegisterICACodeID` (there could definitely be a better naming for it, just an example) — says for itself what kind of ID it returns and hints at the context/reason for that

File: x/interchaintxs/keeper/msg_server.go (hunk 4 above)
Comment (sotnikov-s) [inline x/interchaintxs/keeper/msg_server.go]: why `>=`? we will charge the contracts with src code uploaded before upgrade this way, won't we? seems like here a strict `>` is required

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: app/upgrades/nextupgrade/upgrades_test.go [hunk 8]
```diff
@@ -159,3 +163,56 @@ func (suite *UpgradeTestSuite) TestAdminModuleUpgrade() {
...
+	)
+
+	suite.Require().True(ccvConsumerSubspace.Has(ctx, ccvconsumertypes.KeyRewardDenoms))
+	suite.FundAcc(sdk.AccAddress("neutron1weweewe"), sdk.NewCoins(sdk.NewCoin("untrn", sdk.NewInt(10000))))
+	// emulate mainnet/testnet state
+	ccvConsumerSubspace.Set(ctx, ccvconsumertypes.KeyRewardDenoms, &[]string{params.DefaultDenom})
+
+	var denomsBefore []string
+	ccvConsumerSubspace.Get(ctx, ccvconsumertypes.KeyRewardDenoms, &denomsBefore)
+	suite.Require().Equal(denomsBefore, []string{params.DefaultDenom})
```
Comment (sotnikov-s) [inline app/upgrades/nextupgrade/upgrades_test.go]: you sure it's related to this test case?

File: app/upgrades/nextupgrade/upgrades_test.go [hunk 9]
```diff
@@ -159,3 +163,56 @@ func (suite *UpgradeTestSuite) TestAdminModuleUpgrade() {
...
+	contractKeeper := keeper.NewDefaultPermissionKeeper(app.WasmKeeper)
+	// store contract just to increase code_id
+	_ = suite.StoreTestCode(ctx, sdk.AccAddress("neutron1_ica"), "testdata/neutron_interchain_txs.wasm")
```
Comment (sotnikov-s) [inline app/upgrades/nextupgrade/upgrades_test.go]: why is this needed?

File: app/upgrades/nextupgrade/upgrades_test.go [hunk 10]
```diff
@@ -159,3 +163,56 @@ func (suite *UpgradeTestSuite) TestAdminModuleUpgrade() {
...
+	upgrade := upgradetypes.Plan{
+		Name:   nextupgrade.UpgradeName,
+		Info:   "some text here",
+		Height: 100,
+	}
+	app.UpgradeKeeper.ApplyUpgrade(ctx, upgrade)
+
+	lastCodeID := app.InterchainTxsKeeper.GetLastCodeIDBeforeUpgrade(ctx)
+	// ensure that wasm module stores next code id
+	suite.Require().Equal(lastCodeID, codeIDBefore+1)
```
Comment (sotnikov-s) [inline app/upgrades/nextupgrade/upgrades_test.go:189]: why you expect the next code if it's named the **last** one?

File: x/contractmanager/types/expected_keepers.go (hunk 6 above)
Comment (quasisamurai) [reply x/contractmanager/types/expected_keepers.go:11]: It's not.  This method is required by task itself (get the contract code id by its address)

File: app/upgrades/nextupgrade/upgrades_test.go (hunk 9 above)
Comment (quasisamurai) [reply app/upgrades/nextupgrade/upgrades_test.go]: to emulate that we have more than 1 contract stored before upgrade. Doesn't mandatory, but I find this at least not superfluous

upd. nvm, we already have enough contracts stored. deleted this

File: app/upgrades/nextupgrade/upgrades_test.go (hunk 10 above)
Comment (quasisamurai) [reply app/upgrades/nextupgrade/upgrades_test.go:189]: Because the keyword `last` is  inherited from wasm module. Actually, it's more like `next` id

File: x/interchaintxs/keeper/msg_server.go (hunk 4 above)
Comment (quasisamurai) [reply x/interchaintxs/keeper/msg_server.go]: because as you noticed in the next comment, "_why you expect the next code if it's named the last one?_".  This is related. On a practical level, this value is more like "next id", [here](https://github.com/neutron-org/neutron/blob/7e91506fb6d733163bab7a4a20960558c1674d97/app/upgrades/nextupgrade/upgrades_test.go#L198) it is tested

File: app/upgrades/nextupgrade/upgrades_test.go (hunk 8 above)
Comment (quasisamurai) [reply app/upgrades/nextupgrade/upgrades_test.go]: firstly, I left it because every upgrade test had these lines. now removed, thx

File: x/contractmanager/types/expected_keepers.go (hunk 6 above)
Comment (sotnikov-s) [reply x/contractmanager/types/expected_keepers.go:11]: I can see that you've added this method to two expected wasm keepers: on the interchaintxs module's side and on the contractmanager module's side. the method is used in interchaintxs, but is not in contractmanager
image

File: x/interchaintxs/keeper/msg_server.go [hunk 11]
```diff
@@ -47,8 +47,11 @@ func (k Keeper) RegisterInterchainAccount(goCtx context.Context, msg *ictxtypes.
 		return nil, errors.Wrapf(ictxtypes.ErrNotContract, "%s is not a contract address", msg.FromAddress)
 	}
 
-	if err := k.ChargeFee(ctx, senderAddr, msg.RegisterFee); err != nil {
-		return nil, errors.Wrapf(err, "failed to charge fees to pay for RegisterInterchainAccount msg: %s", msg)
+	// if contract is stored after [last] upgrade, we're not going charge fees for register ICA
+	if k.sudoKeeper.GetContractInfo(ctx, senderAddr).CodeID >= k.GetLastFreeRegisterICACodeID(ctx) {
```
Comment (sotnikov-s) [inline x/interchaintxs/keeper/msg_server.go]: a reminder to improve naming re Last-Current-FromID

**Review by sotnikov-s — APPROVED** [review]

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/6745812174/job/18338461190

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/6745812174
